internal/proto/stdout_stderr_writer.go

```diff
@@ -0,0 +1,133 @@
+package proto
+
+import (
+	"bytes"
+	"io"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"os"
+)
+
+type StdoutStderrWriter interface {
+	GetStdoutWriter() io.Writer
+	GetStderrWriter() io.Writer
+	Close() error
+}
+
+// allows writing of instances of `Output`
+type protoStdoutStderrWriter struct {
+	stdoutReader, stderrReader io.ReadCloser
+	stdoutWriter, stderrWriter io.WriteCloser
+	stdoutDone, stderrDone     chan error
+}
+
+var _ StdoutStderrWriter = protoStdoutStderrWriter{}
+
+func (s protoStdoutStderrWriter) GetStdoutWriter() io.Writer {
+	return s.stdoutWriter
+}
+
+func (s protoStdoutStderrWriter) GetStderrWriter() io.Writer {
+	return s.stderrWriter
+}
+
+func printOutput(output Output) {
+	stdout := output.GetStdout()
+	if stdout != nil {
+		os.Stdout.Write(stdout)
+	}
+
+	stderr := output.GetStderr()
+	if stderr != nil {
+		os.Stderr.Write(stderr)
+	}
+}
+
+// creates a `StdoutStderrWriter` and starts goroutines that feed data written to the writer to
+// the given `sendOutput` function
+func MakeStdoutStderrWriter(sendOutput func(Output)) StdoutStderrWriter {
+	stdoutReader, stdoutWriter := io.Pipe()
+	stderrReader, stderrWriter := io.Pipe()
+	return makeStdoutStderrWriterFromReaderWriters(sendOutput, stdoutReader, stdoutWriter, stderrReader, stderrWriter)
+}
+
+func makeStdoutStderrWriterFromReaderWriters(
+	sendOutput func(Output),
+	stdoutReader io.ReadCloser,
+	stdoutWriter io.WriteCloser,
+	stderrReader io.ReadCloser,
+	stderrWriter io.WriteCloser) StdoutStderrWriter {
+	s := protoStdoutStderrWriter{}
+	s.stdoutReader, s.stdoutWriter, s.stderrReader, s.stderrWriter = stdoutReader, stdoutWriter, stderrReader, stderrWriter
```

Comment:
imo more readable/idiomatic to do as a struct literal:
```
s := protoStdoutStderrWriter{
    stdoutReader: stdoutReader
    ...
}
```

Reply:
yep, much